File: src/main/java/org/moinex/ui/dialog/wallettransaction/RenameWalletController.java
/*
 * Filename: RenameWalletController.java
 * Created on: October  6, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 */

package org.moinex.ui.dialog.wallettransaction;

import jakarta.persistence.EntityExistsException;
import jakarta.persistence.EntityNotFoundException;
import java.util.List;
import javafx.fxml.FXML;
import javafx.scene.control.ComboBox;
import javafx.scene.control.TextField;
import javafx.stage.Stage;
import lombok.NoArgsConstructor;
import org.moinex.model.wallettransaction.Wallet;
import org.moinex.service.I18nService;
import org.moinex.service.WalletService;
import org.moinex.util.Constants;
import org.moinex.util.UIUtils;
import org.moinex.util.WindowUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;

/**
 * Controller for the Rename Wallet dialog
 */
@Controller
@NoArgsConstructor
public class RenameWalletController {
    @FXML private ComboBox<Wallet> walletComboBox;

    @FXML private TextField walletNewNameField;

    private List<Wallet> wallets;

    private WalletService walletService;

    private I18nService i18nService;

    /**
     * Constructor
     * @param walletService WalletService
     * @note This constructor is used for dependency injection
     */
    @Autowired
    public RenameWalletController(WalletService walletService, I18nService i18nService) {
        this.walletService = walletService;
        this.i18nService = i18nService;
    }

    public void setWalletComboBox(Wallet wt) {
        if (wallets.stream().noneMatch(w -> w.getId().equals(wt.getId()))) {
            return;
        }

        walletComboBox.setValue(wt);
    }

    @FXML
    private void initialize() {
        configureComboBoxes();
        loadWalletsFromDatabase();
        populateWalletComboBox();
    }

    @FXML
    private void handleSave() {
        Wallet wt = walletComboBox.getValue();
        String walletNewName = walletNewNameField.getText();

        if (wt == null || walletNewName.isBlank()) {
            WindowUtils.showInformationDialog(
                    i18nService.tr(
                            Constants.TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_TITLE),
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_MESSAGE));
            return;
        }

        try {
            walletService.renameWallet(wt.getId(), walletNewName);
        } catch (IllegalArgumentException | EntityNotFoundException | EntityExistsException e) {
            WindowUtils.showErrorDialog(
                    i18nService.tr(
                            Constants.TranslationKeys
                                    .WALLETTRANSACTION_DIALOG_ERROR_RENAMING_WALLET_TITLE),
                    e.getMessage());
            return;
        }

        Stage stage = (Stage) walletComboBox.getScene().getWindow();
        stage.close();
    }

    @FXML
    private void handleCancel() {
        Stage stage = (Stage) walletComboBox.getScene().getWindow();
        stage.close();
    }

    private void loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName();
    }

    private void populateWalletComboBox() {
        walletComboBox.getItems().addAll(wallets);
    }

    private void configureComboBoxes() {
        UIUtils.configureComboBox(walletComboBox, Wallet::getName);
    }
}

File: src/main/kotlin/org/moinex/ui/dialog/wallettransaction/update/RenameWalletController.kt
/*
 * Filename: RenameWalletController.kt (original filename: RenameWalletController.java)
 * Created on: October  6, 2024
 * Author: Lucas Araújo <araujolucas@dcc.ufmg.br>
 *
 * Migrate to Kotlin on 21/03/2026
 */

package org.moinex.ui.dialog.wallettransaction.update

import jakarta.persistence.EntityExistsException
import jakarta.persistence.EntityNotFoundException
import javafx.fxml.FXML
import javafx.scene.control.ComboBox
import javafx.scene.control.TextField
import javafx.stage.Stage
import org.moinex.common.constants.TranslationKeys
import org.moinex.common.util.UIUtils
import org.moinex.common.util.WindowUtils
import org.moinex.model.wallettransaction.Wallet
import org.moinex.service.PreferencesService
import org.moinex.service.wallet.WalletService
import org.springframework.stereotype.Controller

@Controller
class RenameWalletController(
    private val walletService: WalletService,
    private val preferencesService: PreferencesService,
) {
    @FXML
    private lateinit var walletComboBox: ComboBox<Wallet>

    @FXML
    private lateinit var walletNewNameField: TextField

    private var wallets: List<Wallet> = emptyList()

    fun setWalletComboBox(wt: Wallet) {
        if (wallets.none { it.id == wt.id }) {
            return
        }
        walletComboBox.value = wt
    }

    @FXML
    private fun initialize() {
        configureComboBoxes()
        loadWalletsFromDatabase()
        populateWalletComboBox()
    }

    @FXML
    private fun handleSave() {
        val wt = walletComboBox.value
        val walletNewName = walletNewNameField.text

        if (wt == null || walletNewName.isBlank()) {
            WindowUtils.showInformationDialog(
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_TITLE),
                preferencesService.translate(TranslationKeys.WALLETTRANSACTION_DIALOG_EMPTY_FIELDS_MESSAGE),
            )
            return
        }

        runCatching {
            wt.name = walletNewName
            walletService.renameWallet(wt)
        }.onFailure { e ->
            when (e) {
                is IllegalArgumentException, is EntityNotFoundException, is EntityExistsException -> {
                    WindowUtils.showErrorDialog(
                        preferencesService.translate(
                            TranslationKeys.WALLETTRANSACTION_DIALOG_ERROR_RENAMING_WALLET_TITLE,
                        ),
                        e.message ?: "Unknown error",
                    )
                    return
                }
                else -> throw e
            }
        }

        val stage = walletComboBox.scene.window as Stage
        stage.close()
    }

    @FXML
    private fun handleCancel() {
        val stage = walletComboBox.scene.window as Stage
        stage.close()
    }

    private fun loadWalletsFromDatabase() {
        wallets = walletService.getAllNonArchivedWalletsOrderedByName()
    }

    private fun populateWalletComboBox() {
        walletComboBox.items.addAll(wallets)
    }

    private fun configureComboBoxes() {
        UIUtils.configureComboBox(walletComboBox) { it.name }
    }
}
